patches/tModLoader/Terraria/ModLoader/ChestLoot.cs

```diff
@@ -0,0 +1,321 @@
+	private readonly List<Entry> putCandidates2; // list that contains all entries those going to get from putCandidates
+
+	public IReadOnlyList<Entry> Items {
+		get {
+			List<Entry> entries = items;
+			return entries;
+		}
+	}
+
+	public Entry this[int item] {
+		get {
+			int index = items.FindIndex(x => x.item.type.Equals(item));
+			bool hasInNormal = index != -1;
+			if (hasInNormal)
+				return items[index];
+
+			index = putCandidates2.FindIndex(x => x.item.type.Equals(item));
+			return putCandidates2[index];
+		}
+	}
+
+	public Entry this[Index index] {
+		get {
+			var ind2 = items.ElementAtOrDefault(index);
+			bool hasInNormal = ind2.item != defaultInstance;
+			if (hasInNormal)
+				return ind2;
+
+			return putCandidates2.ElementAtOrDefault(index);
+		}
+	}
+
+	public ChestLoot() {
+		items = new();
+		putCandidates = new();
+		putCandidates2 = new();
+	}
+
+	private void AddCandidates(IList<Entry> entries) {
+		IReadOnlyList<(int nextTo, bool after)> candidates = putCandidates;
+		candidates = candidates.Reverse().ToList();
+
+		List<(int nextTo, bool after)> unsuccessedCandidates = new();
+
+		for (int i = 0; i < 2; i++) { // run twice, cause mods might want to add to chained items
+			var a = candidates;
+			if (i == 1)
+				a = unsuccessedCandidates;
+
+			foreach (var (nextTo, after) in a) {
+				int index = items.FindIndex(x => x.item.type.Equals(nextTo));
+				if (index != -1) {
+					entries.Insert(index + after.ToInt(), putCandidates2[index]);
+				}
+				else {
+					unsuccessedCandidates.Add((nextTo, after));
+				}
+			}
+		}
+	}
+
+	public void AddRange(params Entry[] entries) {
+		foreach (var e in entries) {
+			Add(e);
+		}
+	}
+
+	public void Add(Entry entry) {
+		items.Add(entry);
+	}
+
+	public bool Add(int item, params ICondition[] condition) {
+		return Add(ContentSamples.ItemsByType[item], condition);
+	}
+
+	public bool Add(Item item, params ICondition[] condition) {
+		Add(new Entry(item, condition));
+		return true;
+	}
+
+	private bool PutAt(int destination, Item item, bool after, params ICondition[] condition) {
+		putCandidates.Add(new(destination, after));
+		putCandidates2.Add(new(item, condition));
+		return true;
+	}
+
+	private bool PutAt(int destination, int item, bool after, params ICondition[] condition) {
+		return PutAt(destination, ContentSamples.ItemsByType[item], after, condition);
+	}
+
+	public bool InsertBefore(int destination, int item, params ICondition[] condition) {
+		return PutAt(destination, item, false, condition);
+	}
+
+	public bool InsertBefore(int destination, Item item, params ICondition[] condition) {
+		return PutAt(destination, item, false, condition);
+	}
+
+	public bool InsertAfter(int destination, int item, params ICondition[] condition) {
+		return PutAt(destination + 1, item, true, condition);
+	}
+
+	public bool InsertAfter(int destination, Item item, params ICondition[] condition) {
+		return PutAt(destination + 1, item, true, condition);
+	}
```

Comment:
Is Destination referring to the index of the slot or the target Item to be before/after?

If the latter, please rename to targetItem.
If the former, might need some more work